File: go/dp/decode_ways.go
package main

import (
	"strconv"
)

// 解码方法
// https://leetcode-cn.com/problems/decode-ways

func numDecodings(s string) int {
	dp := make([]int, len(s)+1)
	dp[0] = 1
	if s[0] != '0' {
		dp[1] = 1
	}

	for i := 1; i < len(s); i++ {
		if v, _ := strconv.Atoi(string(s[i])); v > 0 && v < 10 {
			dp[i+1] += dp[i]
		}

		if k, _ := strconv.Atoi(s[i-1 : i+1]); k >= 10 && k <= 26 {
			dp[i+1] += dp[i-1]
		}
	}
	return dp[len(s)]
}

File: python/dp/decode_ways.py

# 解码方法

class Solution:

    def numDecodings(self, s: str) -> int:
        dp = [1] + [0] * len(s)
        if s[0] != '0':
            dp[1] = 1
        
        for i in range(1, len(s)):
            if 0 < int(s[i]) < 10:
                dp[i + 1] += dp[i]
            if 10 <= int(s[i - 1 : i + 1]) <= 26:
                dp[i + 1] += dp[i - 1]
        return dp[-1]
        